### PR TITLE
Treat emmet snippets as textmate

### LINKED ISSUE #19246: Emmet Expansion CSS exit point should be at end of inserted property
Test item #19079

- VSCode Version: Code - Insiders 1.9.0-insider (4265f5e35cb3ea8ba7e0cf6b14ec606b6f8647ea, 2017-01-24T10:29:13.076Z)
- OS Version: Darwin x64 15.6.0
- Extensions: None

Steps to Reproduce:

1. Start with:
```
body {
    bgsz
}
```
2. Expand after `z`
3. Type in some size
4. `<enter>`
5. Expected:
```
body {
    -webkit-background-size: 12px;
    background-size: 12px;<caret>
}
```
6. Actual:
```
body {
    -webkit-background-size: 12px<caret>;
    background-size: 12px;
}
```

Same goes for single line expansions, although there it is not as bad because it is just one arrow-left key to correct.

### PR DESCRIPTION
**Pull request by ramya-rao-a**
Emmet snippets follow text mate syntax.
See https://docs.emmet.io/customization/snippets/#text-snippets
Therefore we should be able to use `CodeSnippet.fromTextmate` instead of `CodeSnippet.fromEmmet` and all the special code we have for emmet

The only catch is when in certain cases emmet returns multiple ${0}
${0} is meant to be the final cursor position and shouldn't occur more than once in a snippet.
This is clearly a bug in emmet. 

_Note_1:_ Emmet 2.0 doesn't have these issues. 

_Note_2:_ There are multiple cases where this happens. I was able to figure out root causes and fixes for a few of these cases, but decided its not worth chasing down these issues in emmet. Mainly because it is in maintenance mode and also, there might be more scenarios which we don't know. We can't fix what we don't know.

Till now, we went around this issue, by assigning [incremental values](https://github.com/Microsoft/vscode/blob/1.11.2/src/vs/editor/contrib/snippet/common/snippet.ts#L263) to such tabstops and [sorting](https://github.com/Microsoft/vscode/blob/1.11.2/src/vs/editor/contrib/snippet/common/snippet.ts#L231) them so that they appear before the numbered placeholder.

So in `<input type="${1:text}" class="${0}" id="${0}" style="${2}">`, the order of tabstops would be
- class
- id
- type
- style

which is not perfect, but this is the current behavior.

In this PR, we replace the multiple ${0} with tab stops starting from the max stop + 1

So `<input type="${1:text}" class="${0}" id="${0}" style="${2}">`
will be `<input type="${1:text}" class="${3}" id="${4}" style="${2}">`

Again, not perfect or ideal, but we don't lose any tabstops which is important.

With this, we can not only fix #19246 but also pave the way for simplifying the way we do snippets
(and remove legacy code, who doesn't like that? :) )

### CODE REVIEW

File: src/vs/editor/contrib/snippet/common/snippet.ts [hunk 1]
```diff
@@ -179,296 +211,296 @@ interface IParsedLine {
...
-			continue;
-		}
-
-		i += 1;
-		convertedSnippet += restOfLine.charAt(0);
-	}
-
-	return convertedSnippet;
-};
+// const InternalFormatSnippetParser = new class implements ISnippetParser {
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: we use source control, no reason push commented code ;-)

File: src/vs/editor/contrib/snippet/common/snippet.ts [hunk 2]
```diff
@@ -55,7 +55,39 @@ export class CodeSnippet implements ICodeSnippet {
...
+		//return EmmetSnippetParser.parse(template);
+
+		let matchFinalStops = template.match(/\$\{0\}/g);
+		if (!matchFinalStops || matchFinalStops.length === 1) {
+			return CodeSnippet.fromTextmate(template);
+		}
+
+		// Emmet sometimes returns snippets with multiple ${0}
+		// In such cases, replace ${0} with incremental tab stops
+
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: There is also the shorthand notion which omits the curlies, like so `foo$1ba$0r`. Isn't emmet using that?

File: src/vs/editor/contrib/snippet/common/snippet.ts [hunk 3]
```diff
@@ -55,7 +55,39 @@ export class CodeSnippet implements ICodeSnippet {
...
+			}
+		}
+		regex = /\$\{(\d+):[^\}]+\}/g;
+		matches = [];
+		while (matches = regex.exec(template)) {
+			let tabStop = Number(matches[1]);
+			if (tabStop > maxTabStop) {
+				maxTabStop = tabStop;
+			}
+		}
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: If this is becoming too complex, e.g `$1`, with escaped `}`, or nested placeholder (stretch, like `foo ${1:\}}` or `foo${2:b${1:ar}$0`), we should consider using the (new) snippet parser (`SnippetParser`) which allows you read all placeholders. Tho I'd say that depends on the Emmet-TextMate-subset

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 1 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: well, you mentioned we shouldn't "delete" the code yet, so I commented it instead :)

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: I haven't seen them being used, but I'll confirm with sergey and get back to this

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 3 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: I'll look into it

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: Emmet always returns ${0} but a user snippet may return $0. 
Updated regex to take care of both

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 3 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: Done

File: src/vs/editor/contrib/snippet/common/snippet.ts (hunk 1 above)
Comment (ramya-rao-a) [inline src/vs/editor/contrib/snippet/common/snippet.ts]: Commented code is now deleted

**Comment by jrieken** [comment]
lgtm

**Pull request merged**